server/src/main/java/io/crate/expression/symbol/DynamicReference.java

```diff
@@ -21,29 +21,36 @@
 
 package io.crate.expression.symbol;
 
-import io.crate.metadata.IndexType;
 import io.crate.metadata.SimpleReference;
 import io.crate.metadata.ReferenceIdent;
 import io.crate.metadata.RowGranularity;
 import io.crate.sql.tree.ColumnPolicy;
 import io.crate.types.DataType;
 import io.crate.types.DataTypes;
 import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
 
 import java.io.IOException;
+import java.util.Objects;
 
 public class DynamicReference extends SimpleReference {
 
-    public DynamicReference(StreamInput in) throws IOException {
-        super(in);
-    }
+    private final ColumnPolicy columnPolicy;
 
-    public DynamicReference(ReferenceIdent ident, RowGranularity granularity, int position) {
+    public DynamicReference(ReferenceIdent ident, RowGranularity granularity, int position, ColumnPolicy columnPolicy) {
         super(ident, granularity, DataTypes.UNDEFINED, position, null);
+        this.columnPolicy = columnPolicy;
     }
 
-    public DynamicReference(ReferenceIdent ident, RowGranularity granularity, ColumnPolicy columnPolicy, int position) {
-        super(ident, granularity, DataTypes.UNDEFINED, columnPolicy, IndexType.PLAIN, true, false, position, null);
+    public DynamicReference(StreamInput in) throws IOException {
+        super(in);
+        columnPolicy = in.readEnum(ColumnPolicy.class);
```

Comment:
This would need version bwc handling,  but can't we derive this from the type?

Reply:
Thank you, adding the version checks. I think the type is fixed to `Undefined` so cannot be derived.